spark/v3.4/spark/src/main/java/org/apache/iceberg/SparkDistributedDataBatchScan.java

```diff
@@ -0,0 +1,332 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.ClosingIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.spark.SparkReadConf;
+import org.apache.iceberg.spark.source.SerializableTableWithSize;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.SparkSession;
+
+public class SparkDistributedDataBatchScan extends DistributedDataBatchScan {
+
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final SparkReadConf readConf;
+
+  private Broadcast<Table> tableBroadcast = null;
+
+  public SparkDistributedDataBatchScan(SparkSession spark, Table table, SparkReadConf readConf) {
+    this(spark, table, readConf, table.schema(), TableScanContext.empty());
+  }
+
+  private SparkDistributedDataBatchScan(
+      SparkSession spark,
+      Table table,
+      SparkReadConf readConf,
+      Schema schema,
+      TableScanContext context) {
+    super(table, schema, context);
+    this.spark = spark;
+    this.sparkContext = JavaSparkContext.fromSparkContext(spark.sparkContext());
+    this.readConf = readConf;
+  }
+
+  @Override
+  protected BatchScan newRefinedScan(
+      Table newTable, Schema newSchema, TableScanContext newContext) {
+    return new SparkDistributedDataBatchScan(spark, newTable, readConf, newSchema, newContext);
+  }
+
+  @Override
+  protected int remoteParallelism() {
+    return sparkContext.defaultParallelism();
+  }
+
+  @Override
+  protected PlanningMode dataPlanningMode() {
+    return readConf.dataPlanningMode();
+  }
+
+  @Override
+  protected List<DataFile> planDataRemotely(List<ManifestFile> manifests) {
+    return sparkContext
+        .parallelize(toBeans(manifests), manifests.size())
+        .flatMap(new ReadDataManifest(tableBroadcast(), context()))
+        .collect();
```

Comment:
Even though this is returning a List, can you have the base method return Iterable so that we can update the implementation later to avoid pulling everything into memory? That way we can also apply additional filters if needed.

Reply:
Changed.